assembly/src/tokens/stream.rs

```diff
@@ -22,63 +22,92 @@ impl<'a> TokenStream<'a> {
+                        || contents.trim().starts_with(Token::EXPORT)
+                        || contents.trim().starts_with(Token::PROC)) =>
+                {
+                    return Err(ParsingError::dangling_procedure_comment(location));
                 }
 
-                // if first dangling comment, then module docs
-                // TODO consider using a dedicated symbol for module docs such as `//!`
+                Some(contents) => contents,
+
+                // first dangling comments are module docs
                 None if tokens.is_empty() => {
-                    module_comment = build_comment(line_info.docs());
+                    module_comment = build_comment(info.docs());
+                    continue;
                 }
 
-                // if has tokens, then dangling docs are illegal
+                // other dangling docs are forbidden
                 None => {
-                    return Err(ParsingError::dangling_procedure_comment(
-                        line_info.line_number() as usize
-                    ));
+                    return Err(ParsingError::dangling_procedure_comment(location));
+                }
+            };
+
+            while !contents.is_empty() {
+                // ignore comments; halt if dangling comment
+                if contents.starts_with(Token::DOC_COMMENT_PREFIX) {
+                    return Err(ParsingError::dangling_procedure_comment(location));
+                } else if contents.starts_with(Token::COMMENT_PREFIX) {
+                    break;
                 }
-            }
 
-            // extend lines until it fits the added tokens
-            lines.resize(tokens.len(), line_number);
+                // fill the doc comments for procedures
+                if contents.starts_with(Token::EXPORT) || contents.starts_with(Token::PROC) {
+                    proc_comments.insert(tokens.len(), build_comment(info.docs()));
+                }
+
+                // pick the current token & remainder
+                let (token, remainder) = match contents.split_once(char::is_whitespace) {
+                    Some(split) => split,
+
+                    // last token; push and break
+                    None => {
+                        tokens.push(contents);
+                        locations.push(location);
+                        break;
+                    }
+                };
+
+                // append the token
+                tokens.push(token);
+                locations.push(location);
+
+                // seek next token
+                let n = match remainder.find(|c: char| !c.is_whitespace()) {
+                    Some(n) => n,
+                    None => break,
+                };
+
+                // update the offset; add extra char consumed by `split_once`
+                location.move_column(token.len() as u32 + n as u32 + 1);
+                contents = remainder.split_at(n).1;
+            }
```

Comment:
I would put this logic into a separate struct - i.e., something like `LineTokenizer`. Also, I think we should be able to introduce this logic with fewer changes to the original code. Specifically, could we replace the the following lines:
```Rust
for token in line.split_whitespace() {
    if token.starts_with(Token::DOC_COMMENT_PREFIX) {
        return Err(ParsingError::dangling_procedure_comment(line_number));
    } else if token.starts_with(Token::COMMENT_PREFIX) {
        break;
    }

    tokens.push(token);
}
```
with something like:
```Rust
let mut line_tokens = LineTokenizer::new(&line_info);
while (token, location) in line_tokens {
    tokens.push(token);
    locations.push(location);
}

if line_tokens.has_dangling_doc_comment() {
    return Err(ParsingError::dangling_procedure_comment(line_number));
}
```

Reply:
`has_dangling_doc_comment` will be somewhat coupled with how `LinesStream` is implemented, that is: the first dangling comment is the module documentation.

The best solution would be to change the syntax from `#!` for procedure comments (Rust equivalent is `///`) to `##`, and leave `#!` for module comments (Rust equivalent is `//!`). We will end up touching many files as all our MASM files are implemented using `#!` for both proc and module docs, but OTOH we wouldn't need to worry about that exception in the future.

I'm pushing a commit with this logic "coupled" inside of `has_dangling_doc_comment` so we can evaluate the result. However, I have the feeling we are creating additional complexity by splitting that responsibility into two structs: `LinesStream` and `LineTokenizer`. My feeling is merging them into one would be a simpler solution.

Reply:
Added on https://github.com/0xPolygonMiden/miden-vm/pull/861/commits/a488bf9a07bc85a4486d7e5844cbeadd8f62ad8e

With that change, `LinesStream` became an implementation detail of `LineTokenizer` so I moved it to the private scope of `assembly/src/tokens/lines.rs`

Reply:
> `has_dangling_doc_comment` will be somewhat coupled with how `LinesStream` is implemented, that is: the first dangling comment is the module documentation.

Why so? `LineTokenizer` does not need to know anything about module comments. The contract is that it gets a line where there is at least one token and then it breaks this line into tokens. The line may optionally end with a comment, and then if the comment is a doc comment (i.e., starts with `#!`) would also set `has_dangling_comment` to true.

> I'm pushing a commit with this logic "coupled" inside of has_dangling_doc_comment so we can evaluate the result. However, I have the feeling we are creating additional complexity by splitting that responsibility into two structs: `LinesStream` and `LineTokenizer`. My feeling is merging them into one would be a simpler solution.

I actually feel like separating them is a cleaner solution - this way we have two components responsible for a small subset of functionality:
- `LineStream` handles splitting the source into lines and extracting relevant doc comments.
- `LineTokenizer` handles splitting a single line of tokens into individual tokens.

We could put handling of dangling line comments into `LineStream` - but I don't think we gain much in terms of clarity there while probably giving up some performance.

Overall, I'd still prefer separating the components the way I described above and actually putting them in separate modules.

Reply:
The intent was to make as close as possible to the suggestion. We, however, cannot use `split_whitespace` as we lose the column there

```rust
let result = ["foo", "bar"].to_vec();
let computed = "foo      bar".split_whitespace().collect::<Vec<_>>();
assert_eq!(result, computed);
```

Instead, we benefit from [char::is_whitespace](https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace) that will handle tabs, or anything unicode for us

Reply:
I mean, the implementation goes for a couple of more lines as we need to have a treatment for some cases like this one.

I added a module split here: https://github.com/0xPolygonMiden/miden-vm/pull/861/commits/16d2532e11829eb0936bc440d06ce3e776901b54